File: bench/cpu/jump.cpp
#include <benchmark/benchmark.h>

template <size_t NopCount>
void Nop() {
  asm("nop");
  if constexpr (NopCount > 1) {
    Nop<NopCount - 1>();
  }
}

template <size_t NopCount>
static void BM_Jump(benchmark::State& state) {
  for (auto _ : state) {
    for (auto i = 0u; i < 100'000; ++i) {
      Nop<NopCount>();
      benchmark::DoNotOptimize(i);
    }
  }
}

BENCHMARK(BM_Jump<1>);
BENCHMARK(BM_Jump<2>);
BENCHMARK(BM_Jump<3>);
BENCHMARK(BM_Jump<4>);
BENCHMARK(BM_Jump<5>);
BENCHMARK(BM_Jump<6>);
BENCHMARK(BM_Jump<7>);
BENCHMARK(BM_Jump<8>);
BENCHMARK(BM_Jump<9>);
BENCHMARK(BM_Jump<10>);
BENCHMARK(BM_Jump<11>);
BENCHMARK(BM_Jump<12>);
BENCHMARK(BM_Jump<13>);
BENCHMARK(BM_Jump<14>);
BENCHMARK(BM_Jump<15>);
BENCHMARK(BM_Jump<16>);
BENCHMARK(BM_Jump<17>);
BENCHMARK(BM_Jump<18>);
BENCHMARK(BM_Jump<19>);
BENCHMARK(BM_Jump<20>);
BENCHMARK(BM_Jump<21>);
BENCHMARK(BM_Jump<22>);
BENCHMARK(BM_Jump<23>);
BENCHMARK(BM_Jump<24>);
BENCHMARK(BM_Jump<25>);
BENCHMARK(BM_Jump<26>);
BENCHMARK(BM_Jump<27>);
BENCHMARK(BM_Jump<28>);
BENCHMARK(BM_Jump<29>);
BENCHMARK(BM_Jump<30>);
BENCHMARK(BM_Jump<31>);
BENCHMARK(BM_Jump<32>);

BENCHMARK_MAIN();

File: bench/cpu/jump.zig
const std = @import("std");

fn printResult(name: []const u8, total_ns: u64, iters: u64) !void {
    const w = std.io.getStdOut().writer();
    try w.print("{s:<40} {d:>8} ns {d:>12}\n", .{ name, total_ns / iters, iters });
}

fn nops(comptime count: usize) void {
    comptime var i = 0;
    inline while (i < count) : (i += 1) {
        asm volatile ("nop");
    }
}

fn benchJump(comptime nop_count: usize) void {
    for (0..100_000) |i| {
        nops(nop_count);
        std.mem.doNotOptimizeAway(i);
    }
}

pub fn main() !void {
    const w = std.io.getStdOut().writer();
    try w.print("{s:<40} {s:>12} {s:>12}\n", .{ "Benchmark", "Time", "Iterations" });
    try w.print("{s}\n", .{"-" ** 66});

    inline for (.{ 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32 }) |n| {
        var total: u64 = 0;
        var iters: u64 = 0;
        while (total < 500_000_000) : (iters += 1) {
            var timer = try std.time.Timer.start();
            benchJump(n);
            total += timer.read();
        }
        try printResult("BM_Jump<" ++ std.fmt.comptimePrint("{}", .{n}) ++ ">", total, iters);
    }
}
